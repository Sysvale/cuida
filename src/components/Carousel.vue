<!-- eslint-disable vue/multi-word-component-names -->
<template>
	<div class="carousel-container">
		<div
			v-if="showArrows"
			class="carousel__arrow carousel__arrow--left"
			:class="{ 'carousel__arrow--dark': darkArrows }"
			@click="scrollToPrevious"
		>
			<icon
				name="arrow-left-outline"
				width="20"
				height="20"
			/>
		</div>

		<div
			ref="carousel"
			v-long-click="startDrag"
			class="carousel"
			@mousedown="startDrag"
			@mousemove="onDrag"
			@mouseup="stopDrag"
			@mouseleave="stopDrag"
			@touchmove="onDrag"
			@touchend="stopDrag"
		>
			<div
				v-for="(item, index) in items"
				:key="item"
				class="carousel__item"
				@click="handleClick(item)"
			>
				<!-- @slot Slot utilizado para renderização de cada item do carrossel. Os dados do escopo do slot podem ser acessados no formato a seguir: ```slot={ item, index }``` -->
				<slot
					:item="item"
					:index="index"
					name="default"
				/>
			</div>
		</div>

		<div
			v-if="showArrows"
			class="carousel__arrow carousel__arrow--right"
			:class="{ 'carousel__arrow--dark': darkArrows }"
			@click="scrollToNext"
		>
			<icon
				name="arrow-right-outline"
				width="20"
				height="20"
			/>
		</div>
	</div>
</template>

<script setup>
import { longClickDirective } from '@sysvale/vue3-long-click';
import useIsMobile from '../utils/composables/useIsMobile';
import { ref, computed } from 'vue';
import Icon from './Icon.vue';

const props = defineProps({
	/**
	 * Array de itens a serem exibidos no carousel.
	 */
	items: {
		type: Array,
		default: () => [],
	},
	/**
	 * Define a posição em que os itens serão alinhados quando for feita a ação de rolagem no carrossel.
	 */
	snapTo: {
		type: String,
		default: 'start',
		validator: (value) => ['start', 'center', 'end'].includes(value),
	},
	/**
	 * Define o espaçamento entre os itens do carrossel.
	 */
	gap: {
		type: Number,
		default: 0,
	},
	/**
	 * Controla a exibição das setas de rolagem.
	 */
	showArrows: {
		type: Boolean,
		default: false,
	},
	/**
	 * Define se a cor das setas de rolagem deve ser escura.
	 */
	darkArrows: {
		type: Boolean,
		default: false,
	},
	/**
	 * Define se os itens do carrossel devem ser clicáveis.
	 */
	clickable: {
		type: Boolean,
		default: false,
	},
});

const emit = defineEmits([
	/**
		* Evento emitido quando algum item do carrossel é clicado.
		* @event item-click
		* @type {Event}
	*/
	'item-click',
]);

const { isMobile } = useIsMobile();
const vLongClick = longClickDirective({ delay: 1000, interval: 100 });
const carousel = ref(null);
let isDragging = false;
let startX, scrollLeft;

const resolvedSnap = computed(() => props.snapTo);
const resolvedGap = computed(() => `${props.gap * 4}px`);

function startDrag(event) {
	if (isMobile.value) {
		return;
	}

	isDragging = true;
	startX = (event.pageX || event.touches[0].pageX) - carousel.value.offsetLeft;
	scrollLeft = carousel.value.scrollLeft;
	event.preventDefault();

	carousel.value.style.scrollSnapType = 'none';
}

function onDrag(event) {
	if (!isDragging) return;
	const x = (event.pageX || event.touches[0].pageX) - carousel.value.offsetLeft;
	const walk = (x - startX) * 1.5;
	carousel.value.scrollLeft = scrollLeft - walk;
}

function stopDrag() {
	isDragging = false;

	carousel.value.style.scrollSnapType = 'x mandatory';
	carousel.value.style.scrollBehavior = 'smooth';

	setTimeout(() => {
		carousel.value.style.scrollBehavior = 'auto';
	}, 300);
}

function scrollToNext() {
	const carouselElement = carousel.value;
	const itemWidth = carouselElement.querySelector('.carousel__item').offsetWidth + props.gap * 4;
	carouselElement.scrollBy({
		left: itemWidth,
		behavior: 'smooth',
	});
}

function scrollToPrevious() {
	const carouselElement = carousel.value;
	const itemWidth = carouselElement.querySelector('.carousel__item').offsetWidth + props.gap * 4;
	carouselElement.scrollBy({
		left: -itemWidth,
		behavior: 'smooth',
	});
}

function handleClick(item) {
	if (!props.clickable) {
		return;
	}

	emit('item-click', item);
}
</script>

<style lang="scss" scoped>
@use '../assets/sass/tokens/index' as tokens;

.carousel-container {
	position: relative;
	width: 100%;
}

.carousel {
	display: flex;
	overflow-x: auto;
	gap: v-bind(resolvedGap);
	scroll-snap-type: x mandatory;
	scrollbar-width: none;
	-ms-overflow-style: none;
	cursor: grab;
	transition: scroll-snap-type 0.3s ease;

	&::-webkit-scrollbar {
		display: none;
	}

	&:active {
		cursor: grabbing;
	}

	&__item {
		flex: none;
		scroll-snap-align: v-bind(resolvedSnap);
	}

	&__arrow {
		position: absolute;
		top: 50%;
		transform: translateY(-50%);
		z-index: 1000;
		background-color: tokens.$n-0;
		color: tokens.$n-700;
		border-radius: 1000px;
		width: 40px;
		height: 40px;
		display: flex;
		justify-content: center;
		align-items: center;
		box-shadow: tokens.$shadow-md;
		cursor: pointer;

		&--left {
			left: 0;
			margin: tokens.ml(3);
		}

		&--right {
			right: 0;
			margin: tokens.mr(3);
		}

		&--dark {
			background-color: tokens.$n-800;
			opacity: 0.85;
			color: tokens.$n-10;
		}
	}

}

</style>