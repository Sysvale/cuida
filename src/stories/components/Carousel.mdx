import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import Carousel from '../../components/Carousel.vue';
import Image from '../../components/Image.vue';
import * as CarouselStories from './Carousel.stories';

<Meta of={CarouselStories} />

export const Template = (args) => ({
	components: { CdsCarousel: Carousel, CdsImage: Image },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-carousel
			v-bind="args"
			v-slot="{ item }"
		>
			<cds-image :src="item" width="300" height="400" />
		</cds-carousel>
	`,
	methods: {},
});

# Carousel

### O Carousel é um componente que permite a exibição de uma série de conteúdos (imagens, textos, cards, etc.) em um formato deslizante, em que o usuário pode navegar entre os itens de forma sequencial.

***

<br />

## Quando usar:

* Quando há necessidade de exibir vários itens (como imagens, produtos, cards informativos) em um espaço reduzido, sem sobrecarregar a interface.
* Para destacar conteúdos importantes de forma dinâmica, como promoções, destaques ou novidades.
* Quando a ordem de exibição dos itens é relevante e a navegação sequencial faz sentido para o contexto.

<br />

## Quando não usar:

* Quando o conteúdo for uma informação crítica e não deve ficar escondida.
* Se a quantidade de itens for muito pequena (menos de 3), pois a navegação pode parecer desnecessária.
* Em interfaces onde a acessibilidade é uma prioridade e o carousel pode dificultar a experiência para usuários com deficiências visuais ou motoras.
* Quando o espaço disponível na tela é insuficiente para exibir os itens de forma clara e legível.

<br />

## Obs:

* Quando o carrossel for utilizado para exibir imagens, é recomendado usar o componente `Image` em vez da tag `<img>` nativa. Isso permite que todas as funcionalidades do <b>Image</b>, como <i>dimmed</i> e <i>opacity</i>, sejam utilizadas.

<br />

## Preview

<Canvas>
  <Story of={CarouselStories.Carousel} />
</Canvas>

<Controls />
