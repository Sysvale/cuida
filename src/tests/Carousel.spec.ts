import { describe, test, expect } from 'vitest';
import Carousel from '../components/Carousel.vue';
import { mount } from '@vue/test-utils';

describe('Carousel', () => {
	test('renders correctly', async () => {
		const wrapper = mount(Carousel, {
			props: {
				items: [
					'https://picsum.photos/600/800?random=1',
					'https://picsum.photos/600/800?random=2',
					'https://picsum.photos/600/800?random=3',
					'https://picsum.photos/600/800?random=4',
					'https://picsum.photos/600/800?random=5',
					'https://picsum.photos/600/800?random=6',
					'https://picsum.photos/600/800?random=7',
				],
				gap: 0,
				snapTo: 'start',
				showArrows: false,
				darkArrows: false,
			},
			slots: {
				default: 'Texto',
			}
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});