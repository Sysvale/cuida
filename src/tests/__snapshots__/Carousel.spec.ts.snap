// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Carousel > renders correctly 1`] = `
"<div data-v-f2893689="" class="carousel-container">
  <!--v-if-->
  <div data-v-f2893689="" class="carousel">
    <div data-v-f2893689="" class="carousel__item">
      <!-- @slot Slot utilizado para renderização de cada item do carrossel. Os dados do escopo do slot podem ser acessados no formato a seguir: \`\`\`slot={ item, index }\`\`\` -->Texto
    </div>
    <div data-v-f2893689="" class="carousel__item">
      <!-- @slot Slot utilizado para renderização de cada item do carrossel. Os dados do escopo do slot podem ser acessados no formato a seguir: \`\`\`slot={ item, index }\`\`\` -->Texto
    </div>
    <div data-v-f2893689="" class="carousel__item">
      <!-- @slot Slot utilizado para renderização de cada item do carrossel. Os dados do escopo do slot podem ser acessados no formato a seguir: \`\`\`slot={ item, index }\`\`\` -->Texto
    </div>
    <div data-v-f2893689="" class="carousel__item">
      <!-- @slot Slot utilizado para renderização de cada item do carrossel. Os dados do escopo do slot podem ser acessados no formato a seguir: \`\`\`slot={ item, index }\`\`\` -->Texto
    </div>
    <div data-v-f2893689="" class="carousel__item">
      <!-- @slot Slot utilizado para renderização de cada item do carrossel. Os dados do escopo do slot podem ser acessados no formato a seguir: \`\`\`slot={ item, index }\`\`\` -->Texto
    </div>
    <div data-v-f2893689="" class="carousel__item">
      <!-- @slot Slot utilizado para renderização de cada item do carrossel. Os dados do escopo do slot podem ser acessados no formato a seguir: \`\`\`slot={ item, index }\`\`\` -->Texto
    </div>
    <div data-v-f2893689="" class="carousel__item">
      <!-- @slot Slot utilizado para renderização de cada item do carrossel. Os dados do escopo do slot podem ser acessados no formato a seguir: \`\`\`slot={ item, index }\`\`\` -->Texto
    </div>
  </div>
  <!--v-if-->
</div>"
`;
